app/models/changeset.rb

```diff
@@ -115,37 +131,43 @@ def trial_succeeds?
   end
 
   def apply!
-    if applied
-      raise Changeset::Error.new(self, 'has already been applied.')
-    else
-      Changeset.transaction do
-        begin
-          change_payloads.each do |change_payload|
-            change_payload.apply!
-          end
-          self.update(applied: true, applied_at: Time.now)
-          # Destroy change payloads
-          change_payloads.destroy_all
-        rescue
-          logger.error "Error applying Changeset #{self.id}: #{$!.message}"
-          logger.error $!.backtrace
-          raise Changeset::Error.new(self, $!.message, $!.backtrace)
+    fail raise Changeset::Error.new(self, 'has already been applied.') if applied
+    Changeset.transaction do
+      begin
+        change_payloads.each do |change_payload|
+          change_payload.apply!
         end
-      end
-      unless Figaro.env.send_changeset_emails_to_users.presence == 'false'
-        if self.user && self.user.email.present? && !self.user.admin
-          ChangesetMailer.delay.application(self.id)
+        self.update(applied: true, applied_at: Time.now)
+        # Create any feed-entity associations
+        if feed && feed_version
+          self.entities_created_or_updated.each do |entity|
```

Comment:
I'd recommend `.find_each` instead of `.each`, so not as many models are loaded in to memory in one go.

Another option: Why not use `find_in_batches` and then bulk insert batches of EIFF models using activerecord-import?